pkg/kubectl/cmd/drain.go

```diff
@@ -0,0 +1,343 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/kubelet/types"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+type DrainOptions struct {
+	client             *client.Client
+	factory            *cmdutil.Factory
+	Force              bool
+	GracePeriodSeconds int
+	mapper             meta.RESTMapper
```

Comment:
> DrainOptions should contain only Force and GracePeriodSeconds as it refers to command's options. Other fields as client, factory, mapper, etc. are related to working with resources. What about to rename the struct into DrainConfig struct? This way it is clear all fields are configuration related.

We are following this pattern elsewhere, so if we decide on a name it should be the same in all places. The rest of the fields are used so that parts like validation and run of the command won't actually need to access cobra, factory, or other machinery. If you have opinions on this, please comment in https://github.com/kubernetes/kubernetes/pull/17366.